File: experiments/lemonade-wars/src/lemonwars/customers.py
"""Customer demand and the daily sales resolution.

Demand is driven by the district's demographic mix, weather, season, player
reputation, advertising awareness, competition, tier capacity, technology and
active events. Each simulated customer then weighs the price against a
demographic price-sensitivity before buying; every purchase produces a
satisfaction score that feeds reputation and repeat business.
"""

from __future__ import annotations

import math
import random
from dataclasses import dataclass, field

from . import data


# --------------------------------------------------------------------------
# District / demographic fit
# --------------------------------------------------------------------------

def district_ideal(game, include_weather: bool = True) -> dict[str, float]:
    """The flavour vector this district most wants right now.

    Demographic tastes are mixed by their share of the district, then nudged
    by weather and season (hot days pull toward refreshment, cold days toward
    warmth).
    """
    d = data.DISTRICTS[game.player.district]
    ideal = {dim: 0.0 for dim in data.FLAVOUR_DIMS}
    for demog_key, share in d.mix.items():
        taste = data.DEMOGRAPHICS[demog_key].taste
        for dim in data.FLAVOUR_DIMS:
            ideal[dim] += taste[dim] * share
    if include_weather:
        w = game.world.weather_at(game.player.district)
        season = data.season_for_day(game.day)
        shift = (w.refresh_shift + season.refresh_shift) * game.difficulty.weather_mult
        ideal["refresh"] += shift
        shift_w = (w.warm_shift + season.warm_shift) * game.difficulty.weather_mult
        ideal["warm"] += shift_w
        ideal["fresh"] -= max(0, shift) * 0.3
    return {dim: min(max(v, 0), 100) for dim, v in ideal.items()}


def product_fit(game, flavour: dict, demog_key: str | None = None,
                include_weather: bool = True) -> float:
    """0..100 — how well a product's flavour matches its market."""
    if demog_key is not None:
        demog = data.DEMOGRAPHICS[demog_key]
        dist = 0.0
        wsum = 0.0
        for dim in data.FLAVOUR_DIMS:
            w = demog.pref[dim]
            dist += w * abs(flavour[dim] - demog.taste[dim])
            wsum += w
        return round(100.0 - dist / max(wsum, 1e-6), 1)
    ideal = district_ideal(game, include_weather)
    d = data.DISTRICTS[game.player.district]
    dist = 0.0
    wsum = 0.0
    for dim in data.FLAVOUR_DIMS:
        w = 0.0
        for demog_key, share in d.mix.items():
            w += data.DEMOGRAPHICS[demog_key].pref[dim] * share
        dist += w * abs(flavour[dim] - ideal[dim])
        wsum += w
    return round(100.0 - dist / max(wsum, 1e-6), 1)


# --------------------------------------------------------------------------
# Demand
# --------------------------------------------------------------------------

def expected_customers(game) -> float:
    """Raw expected foot traffic before stock limits."""
    p = game.player
    d = data.DISTRICTS[p.district]
    season = data.season_for_day(game.day)
    w = game.world.weather_at(p.district)

    base = d.demand * season.foot * w.foot * p.tier().cust_mult

    rep_mult = 0.6 + p.reputation / 160.0
    awareness_mult = 0.65 + 0.35 * (1.0 - math.exp(-p.awareness / 120.0))
    comp_mult = 1.0 - d.competition * 0.30 * (0.6 + 0.4 * game.difficulty.competitor_mult)
    tech_mult = 1.0
    if "online_ordering" in p.tech:
        tech_mult *= 1.15
    if "delivery_app" in p.tech:
        tech_mult *= 1.10
    storefront = 1.4 if "storefront" in p.equipment else 1.0

    flavour = game.recipe.flavour(data.PRODUCTS[game.active_product])
    fit_mult = 0.7 + 0.6 * product_fit(game, flavour) / 100.0

    event_mult = 1.0
    for name, (mult, _days) in game.demand_modifiers.items():
        event_mult *= mult

    cap = p.tier().max_serve
    if "cashier" in p.employees:
        cap *= 1.10
    if "cash_register" in p.equipment:
        cap *= 1.15

    customers = base * rep_mult * awareness_mult * comp_mult * tech_mult * \
        storefront * fit_mult * event_mult
    return min(max(customers, 5.0), cap)


# --------------------------------------------------------------------------
# Sales resolution
# --------------------------------------------------------------------------

@dataclass
class SalesReport:
    customers: int = 0
    sold: int = 0
    lost: int = 0
    revenue: float = 0.0
    avg_satisfaction: float = 0.0
    repeats: int = 0
    quality: float = 0.0
    reference: float = 0.0
    fit: float = 0.0
    weather_fit: float = 0.0
    satisfaction_samples: list = field(default_factory=list)


def weather_fit_for(game, temp: str) -> float:
    """How well the serving temperature matches today's weather."""
    w = game.world.weather_at(game.player.district)
    hot_drink = temp in ("hot", "warm")
    if not hot_drink:
        if w.key in ("heat_wave", "sunny", "humid"):
            return 16.0
        if w.key in ("cold_snap", "snow"):
            return -12.0
        if w.key in ("rain", "thunderstorm", "windy"):
            return -5.0
    else:
        if w.key in ("cold_snap", "snow", "rain", "thunderstorm"):
            return 16.0
        if w.key in ("heat_wave", "sunny", "humid"):
            return -14.0
    return 0.0


def reference_price(game) -> float:
    product = data.PRODUCTS[game.active_product]
    d = data.DISTRICTS[game.player.district]
    return round(product.anchor * d.cost_mult * (1.0 + game.player.reputation / 600.0), 2)


def run_sales(game) -> SalesReport:
    p = game.player
    report = SalesReport()
    product = data.PRODUCTS[game.active_product]
    stock = p.product_stock(game.active_product)
    if stock <= 0.01:
        report.customers = int(expected_customers(game))
        report.lost = report.customers
        return report

    flavour = game.recipe.flavour(product)
    quality = game.recipe.quality(game, product, flavour)
    ref = reference_price(game)
    ratio = game.price / max(ref, 0.01)
    temp = game.recipe.temp
    wfit = weather_fit_for(game, temp)

    report.quality = quality
    report.reference = ref
    report.weather_fit = wfit
    report.fit = product_fit(game, flavour)

    demog_keys = list(data.DEMOGRAPHICS)
    d = data.DISTRICTS[p.district]
    mix_weights = [d.mix.get(k, 0.0) for k in demog_keys]
    patience = game.difficulty.patience
    loyalty = 1.2 if "loyalty" in p.tech else 1.0
    revenue_mult = 1.05 if "predictive_pricing" in p.tech else 1.0
    service_clean = 45.0 + p.staff_bonus("service") + p.staff_bonus("cleanliness") * 0.5
    cleanliness = min(100.0, p.cleanliness + p.staff_bonus("cleanliness"))

    total_sat = 0.0
    for _ in range(int(expected_customers(game))):
        if stock <= 0.01:
            report.lost += 1
            continue
        demog = data.DEMOGRAPHICS[game.rng.choices(demog_keys, weights=mix_weights)[0]]
        price_sens = demog.price_sens * patience
        buy_chance = 0.92 - price_sens * max(0.0, ratio - 1.0) * 1.5 \
            - (1.0 - quality / 100.0) * 0.22
        buy_chance = min(max(buy_chance, 0.04), 0.98)
        if game.rng.random() >= buy_chance:
            continue

        stock -= 1
        report.sold += 1
        sat = 0.58 * quality + 0.10 * product_fit(game, flavour, demog.key) \
            + 0.14 * wfit + 0.10 * min(100.0, service_clean) \
            - max(0.0, (ratio - 1.0) * price_sens * 40.0)
        sat = min(max(sat, 3.0), 100.0)
        report.satisfaction_samples.append(sat)
        total_sat += sat

        if game.rng.random() < (0.04 + 0.5 * sat / 100.0) * loyalty:
            report.repeats += 1
        price_now = game.price * revenue_mult
        report.revenue += price_now
        p.sales_tax_bill = round(p.sales_tax_bill + price_now * 0.08, 2)

    p.products[game.active_product]["qty"] = round(stock, 3)
    if stock <= 0.01:
        p.products.pop(game.active_product, None)

    report.customers = report.sold + report.lost
    if report.satisfaction_samples:
        report.avg_satisfaction = round(sum(report.satisfaction_samples) / len(report.satisfaction_samples), 1)
        delta = min(max((report.avg_satisfaction - 58.0) * 0.015, -2.0), 3.0)
        p.reputation = round(min(max(p.reputation + delta, 0.0), 100.0), 1)
    if report.lost > 0:
        p.reputation = round(max(0.0, p.reputation - min(3.0, report.lost / 25.0)), 1)

    # stats
    game.stats["customers"] += report.customers
    game.stats["sold"] += report.sold
    game.stats["lost_sales"] += report.lost
    game.stats["repeat_customers"] += report.repeats
    game.stats["revenue"] = round(game.stats["revenue"] + report.revenue, 2)
    p.cash = round(p.cash + report.revenue, 2)
    return report

File: experiments/lemonade-wars/js/customers.js
// Customer demand and daily sales resolution — direct port of customers.py.
"use strict";

import { DEMOGRAPHICS, DISTRICTS, PRODUCTS, season_for_day } from "./data.js";

// --------------------------------------------------------------------------
// District / demographic fit
// --------------------------------------------------------------------------

export function districtIdeal(game, includeWeather = true) {
  const d = DISTRICTS[game.player.district];
  const ideal = {};
  for (const dim of Object.keys(DEMOGRAPHICS[Object.keys(d.mix)[0]].taste)) ideal[dim] = 0.0;
  for (const [demogKey, share] of Object.entries(d.mix)) {
    const taste = DEMOGRAPHICS[demogKey].taste;
    for (const dim of Object.keys(ideal)) ideal[dim] += taste[dim] * share;
  }
  if (includeWeather) {
    const w = game.world.weatherAt(game.player.district);
    const season = season_for_day(game.day);
    const shift = (w.refresh_shift + season.refresh_shift) * game.difficulty.weather_mult;
    ideal.refresh += shift;
    const shiftW = (w.warm_shift + season.warm_shift) * game.difficulty.weather_mult;
    ideal.warm += shiftW;
    ideal.fresh -= Math.max(0, shift) * 0.3;
  }
  const out = {};
  for (const [dim, v] of Object.entries(ideal)) out[dim] = clamp(v, 0, 100);
  return out;
}

export function productFit(game, flavour, demogKey = null, includeWeather = true) {
  if (demogKey !== null) {
    const demog = DEMOGRAPHICS[demogKey];
    let dist = 0.0;
    let wsum = 0.0;
    for (const dim of Object.keys(flavour)) {
      const w = demog.pref[dim];
      dist += w * Math.abs(flavour[dim] - demog.taste[dim]);
      wsum += w;
    }
    return round1(100.0 - dist / Math.max(wsum, 1e-6));
  }
  const ideal = districtIdeal(game, includeWeather);
  const d = DISTRICTS[game.player.district];
  let dist = 0.0;
  let wsum = 0.0;
  for (const dim of Object.keys(flavour)) {
    let w = 0.0;
    for (const [demogKey, share] of Object.entries(d.mix)) {
      w += DEMOGRAPHICS[demogKey].pref[dim] * share;
    }
    dist += w * Math.abs(flavour[dim] - ideal[dim]);
    wsum += w;
  }
  return round1(100.0 - dist / Math.max(wsum, 1e-6));
}

// --------------------------------------------------------------------------
// Demand
// --------------------------------------------------------------------------

export function expectedCustomers(game) {
  const p = game.player;
  const d = DISTRICTS[p.district];
  const season = season_for_day(game.day);
  const w = game.world.weatherAt(p.district);

  const base = d.demand * season.foot * w.foot * p.tier().cust_mult;

  const repMult = 0.6 + p.reputation / 160.0;
  const awarenessMult = 0.65 + 0.35 * (1.0 - Math.exp(-p.awareness / 120.0));
  const compMult = 1.0 - d.competition * 0.30 * (0.6 + 0.4 * game.difficulty.competitor_mult);
  let techMult = 1.0;
  if (p.tech.has("online_ordering")) techMult *= 1.15;
  if (p.tech.has("delivery_app")) techMult *= 1.10;
  const storefront = p.equipment.has("storefront") ? 1.4 : 1.0;

  const flavour = game.recipe.flavour(PRODUCTS[game.active_product]);
  const fitMult = 0.7 + (0.6 * productFit(game, flavour)) / 100.0;

  let eventMult = 1.0;
  for (const [name, [mult]] of Object.entries(game.demand_modifiers)) eventMult *= mult;

  let cap = p.tier().max_serve;
  if (p.employees.cashier) cap *= 1.10;
  if (p.equipment.has("cash_register")) cap *= 1.15;

  const customers = base * repMult * awarenessMult * compMult * techMult * storefront * fitMult * eventMult;
  return Math.min(Math.max(customers, 5.0), cap);
}

// --------------------------------------------------------------------------
// Sales resolution
// --------------------------------------------------------------------------

export class SalesReport {
  constructor() {
    this.customers = 0;
    this.sold = 0;
    this.lost = 0;
    this.revenue = 0.0;
    this.avg_satisfaction = 0.0;
    this.repeats = 0;
    this.quality = 0.0;
    this.reference = 0.0;
    this.fit = 0.0;
    this.weather_fit = 0.0;
    this.satisfaction_samples = [];
  }
}

export function weatherFitFor(game, temp) {
  const w = game.world.weatherAt(game.player.district);
  const hotDrink = temp === "hot" || temp === "warm";
  if (!hotDrink) {
    if (w.key === "heat_wave" || w.key === "sunny" || w.key === "humid") return 16.0;
    if (w.key === "cold_snap" || w.key === "snow") return -12.0;
    if (w.key === "rain" || w.key === "thunderstorm" || w.key === "windy") return -5.0;
  } else {
    if (w.key === "cold_snap" || w.key === "snow" || w.key === "rain" || w.key === "thunderstorm") return 16.0;
    if (w.key === "heat_wave" || w.key === "sunny" || w.key === "humid") return -14.0;
  }
  return 0.0;
}

export function referencePrice(game) {
  const product = PRODUCTS[game.active_product];
  const d = DISTRICTS[game.player.district];
  return round2(product.anchor * d.cost_mult * (1.0 + game.player.reputation / 600.0));
}

export function runSales(game) {
  const p = game.player;
  const report = new SalesReport();
  const product = PRODUCTS[game.active_product];
  let stock = p.productStock(game.active_product);
  if (stock <= 0.01) {
    report.customers = Math.floor(expectedCustomers(game));
    report.lost = report.customers;
    return report;
  }

  const flavour = game.recipe.flavour(product);
  const quality = game.recipe.quality(game, product, flavour);
  const ref = referencePrice(game);
  const ratio = game.price / Math.max(ref, 0.01);
  const temp = game.recipe.temp;
  const wfit = weatherFitFor(game, temp);

  report.quality = quality;
  report.reference = ref;
  report.weather_fit = wfit;
  report.fit = productFit(game, flavour);

  const demogKeys = Object.keys(DEMOGRAPHICS);
  const d = DISTRICTS[p.district];
  const mixWeights = demogKeys.map((k) => d.mix[k] ?? 0.0);
  const patience = game.difficulty.patience;
  const loyalty = p.tech.has("loyalty") ? 1.2 : 1.0;
  const revenueMult = p.tech.has("predictive_pricing") ? 1.05 : 1.0;
  const serviceClean = 45.0 + p.staffBonus("service") + p.staffBonus("cleanliness") * 0.5;
  const cleanliness = Math.min(100.0, p.cleanliness + p.staffBonus("cleanliness"));

  let totalSat = 0.0;
  const nCustomers = Math.floor(expectedCustomers(game));
  for (let i = 0; i < nCustomers; i++) {
    if (stock <= 0.01) {
      report.lost += 1;
      continue;
    }
    const demog = DEMOGRAPHICS[game.rng.choices(demogKeys, mixWeights)];
    const priceSens = demog.price_sens * patience;
    let buyChance = 0.92 - priceSens * Math.max(0.0, ratio - 1.0) * 1.5 - (1.0 - quality / 100.0) * 0.22;
    buyChance = clamp(buyChance, 0.04, 0.98);
    if (game.rng.random() >= buyChance) continue;

    stock -= 1;
    report.sold += 1;
    let sat =
      0.58 * quality +
      0.10 * productFit(game, flavour, demog.key) +
      0.14 * wfit +
      0.10 * Math.min(100.0, serviceClean) -
      Math.max(0.0, (ratio - 1.0) * priceSens * 40.0);
    sat = clamp(sat, 3.0, 100.0);
    report.satisfaction_samples.push(sat);
    totalSat += sat;

    if (game.rng.random() < (0.04 + (0.5 * sat) / 100.0) * loyalty) report.repeats += 1;
    const priceNow = game.price * revenueMult;
    report.revenue += priceNow;
    p.sales_tax_bill = round2(p.sales_tax_bill + priceNow * 0.08);
  }

  p.products[game.active_product].qty = round3(stock);
  if (stock <= 0.01) delete p.products[game.active_product];

  report.customers = report.sold + report.lost;
  if (report.satisfaction_samples.length > 0) {
    report.avg_satisfaction = round1(totalSat / report.satisfaction_samples.length);
    const delta = clamp((report.avg_satisfaction - 58.0) * 0.015, -2.0, 3.0);
    p.reputation = round1(clamp(p.reputation + delta, 0.0, 100.0));
  }
  if (report.lost > 0) {
    p.reputation = round1(Math.max(0.0, p.reputation - Math.min(3.0, report.lost / 25.0)));
  }

  // stats
  game.stats.customers += report.customers;
  game.stats.sold += report.sold;
  game.stats.lost_sales += report.lost;
  game.stats.repeat_customers += report.repeats;
  game.stats.revenue = round2(game.stats.revenue + report.revenue);
  p.cash = round2(p.cash + report.revenue);
  return report;
}

const round1 = (v) => Math.round(v * 10) / 10;
const round2 = (v) => Math.round(v * 100) / 100;
const round3 = (v) => Math.round(v * 1000) / 1000;
const clamp = (v, lo, hi) => Math.min(Math.max(v, lo), hi);
